File: Construct_Binary_Tree_from_Preorder_and_Inorder.py
# Created by Aashish Adhikari at 11:36 AM 1/15/2021

'''
Time Complexity:
If k is the average length of the number of elements to each node, O(n . k ) since at each node we are calculating the index of the root element.

Space Complexity:
O(n)


'''



# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def helper(self, preorder, inorder):

        # base case
        if len(preorder) == 0:
            return None


        # logic
        root = preorder[0]
        idx_of_root = inorder.index(root)

        left_subtree_elements_inorder = inorder[:idx_of_root]
        right_subtree_elements_inorder = inorder[idx_of_root+1 :]

        num_left_subtree_elements = len(left_subtree_elements_inorder)
        num_right_subtree_elements = len(right_subtree_elements_inorder)

        root_node = TreeNode(root)
        root_node.left = self.helper(preorder[1:num_left_subtree_elements+1] , left_subtree_elements_inorder)
        root_node.right = self.helper(preorder[num_left_subtree_elements+1:] , right_subtree_elements_inorder)

        return  root_node


    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """

        return self.helper(preorder, inorder)

File: ConstructBinaryTreeFromPreorderAndInorder.java
public class ConstructBinaryTreeFromPreorderAndInorder {
    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     *     int val;
     *     TreeNode left;
     *     TreeNode right;
     *     TreeNode() {}
     *     TreeNode(int val) { this.val = val; }
     *     TreeNode(int val, TreeNode left, TreeNode right) {
     *         this.val = val;
     *         this.left = left;
     *         this.right = right;
     *     }
     * }
     */
    Map<Integer, Integer> inorderMap = new HashMap<>();

    // TC: O(N) where N is number of nodes
    // SC: O(N + H) where N is max height of tree and H is max height of tree
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for (int i = 0; i < inorder.length; i++) {
            inorderMap.put(inorder[i], i);
        }
        return helper(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }

    private TreeNode helper(int[] preorder, int preStart, int preEnd, int[] inorder, int inStart, int inEnd) {
        if (preStart > preEnd || inStart > inEnd) {
            return null;
        }
        int rootVal = preorder[preStart];
        int rootIndex = inorderMap.get(rootVal);
        int leftTreeLen = rootIndex - inStart;

        TreeNode root = new TreeNode(rootVal);
        root.left = helper(preorder, preStart + 1, preStart + leftTreeLen, inorder, inStart, rootIndex - 1);
        root.right = helper(preorder, preStart + leftTreeLen + 1, preEnd, inorder, rootIndex + 1, inEnd);
        return root;
    }
}
